### PR TITLE
Support removing remote applications

### PR DESCRIPTION
**Pull request by wallyworld**
Remote applications can now be removed using:
$ juju remove-application <remote-app-name>

Also include some drive by service->application renames.

QA:
bootstrap juju
add remote application a relation
remove remote application
check that remote application and relation is removed

### CODE REVIEW

File: cmd/juju/application/removeapplication_test.go [hunk 1]
```diff
@@ -43,70 +44,88 @@ func (s *RemoveServiceSuite) SetUpTest(c *gc.C) {
...
+	err := runRemoveApplication(c, "riak")
 	c.Assert(err, jc.ErrorIsNil)
 	riak, err := s.State.Application("riak")
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(riak.Life(), gc.Equals, state.Dying)
 	s.stub.CheckNoCalls(c)
 }
 
-func (s *RemoveServiceSuite) TestRemoveLocalMetered(c *gc.C) {
+func (s *RemoveApplicationSuite) TestRemoteApplication(c *gc.C) {
```
Comment (mjs) [inline cmd/juju/application/removeapplication_test.go]: nit: given the suite already has Application in the name you don't need it again in the test name.

File: state/cleanup.go [hunk 2]
```diff
@@ -215,16 +215,41 @@ func (st *State) cleanupMachinesForDyingModel() (err error) {
...
 	applications, closer := st.getCollection(applicationsC)
 	defer closer()
 	application := Application{st: st}
 	sel := bson.D{{"life", Alive}}
 	iter := applications.Find(sel).Iter()
-	defer closeIter(iter, &err, "reading service document")
+	defer closeIter(iter, &err, "reading application document")
+	for iter.Next(&application.doc) {
+		if err := application.Destroy(); err != nil {
+			return err
```
Comment (mjs) [inline state/cleanup.go]: trace?

File: state/cleanup.go [hunk 3]
```diff
@@ -215,16 +215,41 @@ func (st *State) cleanupMachinesForDyingModel() (err error) {
...
+	// via individual transactions, because they could be in any state at all.
+	remoteApp, closer := st.getCollection(remoteApplicationsC)
+	defer closer()
+	application := RemoteApplication{st: st}
+	sel := bson.D{{"life", Alive}}
+	iter := remoteApp.Find(sel).Iter()
+	defer closeIter(iter, &err, "reading remote application document")
 	for iter.Next(&application.doc) {
 		if err := application.Destroy(); err != nil {
 			return err
```
Comment (mjs) [inline state/cleanup.go]: trace?

File: state/cleanup.go [hunk 4]
```diff
@@ -215,16 +215,41 @@ func (st *State) cleanupMachinesForDyingModel() (err error) {
...
+}
+
+func (st *State) removeRemoteApplicationsForDyingModel() (err error) {
+	// This won't miss remote applications, because a Dying model cannot have
+	// applications added to it. But we do have to remove the applications themselves
+	// via individual transactions, because they could be in any state at all.
+	remoteApp, closer := st.getCollection(remoteApplicationsC)
+	defer closer()
+	application := RemoteApplication{st: st}
+	sel := bson.D{{"life", Alive}}
```
Comment (mjs) [inline state/cleanup.go]: extract all the code below out into a func? it's the same for the local and remote case.

File: cmd/juju/application/removeapplication_test.go (hunk 1 above)
Comment (wallyworld) [inline cmd/juju/application/removeapplication_test.go]: the suite also contains tests for adding remote relations against a remote app ie TestAddRemoteRelation() so the test name is ok I think

File: state/cleanup.go (hunk 4 above)
Comment (wallyworld) [inline state/cleanup.go]: it's not really the same - one uses a RemoteApplication struct, the other an Application struct. I've renamed some variables to make it more obvious the code is different

File: state/cleanup.go (hunk 2 above)
Comment (wallyworld) [inline state/cleanup.go]: fixed

File: state/cleanup.go (hunk 3 above)
Comment (wallyworld) [inline state/cleanup.go]: fixed

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**